### DISCUSSION TITLE
Large Twisted TMD Bilayer System

### DISCUSSION THREAD

**Discussion by jdgeorga** [ttadano/alamode#114] [Q&A] [answered]
Hello, 

I'm looking to calculate the phonon-phonon interactions in a large twisted bilayer TMD system. My system has 42 atoms in the primitive cell, and I'm only looking at a 1x1x1 supercell. I'm running the ALM code to calculate the 2nd and 3rd order force constants. To minimize the number of displacements (which are >100000 for my system), I am using a cutoff of 10 for each atom, and I am using random displacements + the LASSO method described in the ALAMODE paper. This method works great on a smaller system monolayer system (3 atoms), and even slightly larger (12 atoms), but gets stuck in the CONSTRAINTS section of the code for the 42 atom system. _Note that I have increased the number of OMP_THREADS to 114_

Specifically this section of the output and immediately afterwards takes the longest (> 2400 seconds):

> Generating constraints from crystal symmetry
>   in crystallographic (fractional) coordinates ...
>    HARMONIC done.
>     ANHARM3 done.
>   Finished !
> 
>   Generating constraints for translational invariance ...
>    HARMONIC ... done.
>     ANHARM3 ... done.
>   Finished !

Looking into the running processes is showing that during this section there is no parallelization with OMP_THREADS, and might simply be a serial process. 

Is there some way to speed this section up?

Thanks,

Johnathan

**Comment by ttadano** [comment, marked as answer]
The ALM code generates all constraints between force constants (symmetry, translation, rotation) and merges them to make a single constraint matrix in rref (reduced row echelon form). This rref reduction is not parallelized and is likely the bottleneck point. Unfortunately, this is an algorithmic limitation and not easy to solve.

One possible way to improve the performance is to set `FCSYM_BASIS = Cartesian` (https://alamode.readthedocs.io/en/latest/almdir/formalism_alm.html#constraints-between-ifcs).

**Reply by jdgeorga** [reply]
Thanks for the quick response. Setting the symmetry basis to Cartesian reduced the run time by about 25%, but the larger system are still intractable. **Is it possible to skip the constraint step entirely at the cost of needing to fit more FCs?** 

My thinking is that I can cheaply calculate forces with LAMMPS (MD) and I'm not limited by OpenMP threads or processors but rather the size of the system. And my understanding is that the fitting step can be parallelized fairly well.
